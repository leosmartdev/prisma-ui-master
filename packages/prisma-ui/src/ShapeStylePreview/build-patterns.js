// import { __ } from './i18n';

const SIZE = 64;
const SPACING = 16;
const STROKEWIDTH = 2;
const SYMBOLSIZE = 8;

const makeCanvas = () => {
  const canvas = document.createElement('canvas');
  canvas.width = SIZE;
  canvas.height = SIZE;
  return { canvas, g: canvas.getContext('2d') };
};

const makePattern = (canvas, g) => g.createPattern(canvas, 'repeat');

/**
 * Creates a solid background of color
 * @param {object} style
 * @param {string} style.fill The color.
 */
export const solid = (style) => {
  const { canvas, g } = makeCanvas();
  g.fillStyle = style.fill;
  g.fillRect(0, 0, SIZE, SIZE);
  return makePattern(canvas, g);
};

/**
 * Creates a horizontal pattern of lines.
 * |--------------|
 * |--------------|
 * |--------------|
 * |--------------|
 * |--------------|
 * |--------------|
 * @param {object} style
 * @param {string} style.fill Background color for the fill pattern.
 * @param {string} style.stroke Color for the fill pattern stroke lines.
 */
export const horizontal = (style) => {
  const { canvas, g } = makeCanvas();
  g.fillStyle = style.fill;

  g.fillRect(0, 0, SIZE, SIZE);
  g.strokeStyle = style.stroke;
  g.lineWidth = STROKEWIDTH;
  g.beginPath();
  for (let i = 0; i <= SIZE; i += SPACING) {
    g.moveTo(0, i);
    g.lineTo(SIZE, i);
  }
  g.stroke();
  return makePattern(canvas, g);
};

/**
 * Creates a vertical pattern of lines.
 * ---------------
 * | | | | | | | |
 * | | | | | | | |
 * | | | | | | | |
 * | | | | | | | |
 * ---------------
 * @param {object} style
 * @param {string} style.fill Background color for the fill pattern.
 * @param {string} style.stroke Color for the fill pattern stroke lines.
 */
export const vertical = (style) => {
  const { canvas, g } = makeCanvas();
  g.fillStyle = style.fill;
  g.fillRect(0, 0, SIZE, SIZE);

  g.strokeStyle = style.stroke;
  g.lineWidth = STROKEWIDTH;
  g.beginPath();
  for (let i = 0; i <= SIZE; i += SPACING) {
    g.moveTo(i, 0);
    g.lineTo(i, SIZE);
  }
  g.stroke();
  return makePattern(canvas, g);
};

/**
 * Creates a pattern of slashes.
 * ---------------
 * | / / / / / / |
 * |/ / / / / / /|
 * | / / / / / / |
 * |/ / / / / / /|
 * ---------------
 * @param {object} style
 * @param {string} style.fill Background color for the fill pattern.
 * @param {string} style.stroke Color for the fill pattern stroke lines.
 */
export const slash = (style) => {
  const { canvas, g } = makeCanvas();
  g.fillStyle = style.fill;
  g.fillRect(0, 0, SIZE, SIZE);

  g.strokeStyle = style.stroke;
  g.lineWidth = STROKEWIDTH;
  g.beginPath();
  for (let i = 0; i <= SIZE; i += SPACING) {
    g.moveTo(0, i);
    g.lineTo(i, 0);
    g.moveTo(i, SIZE);
    g.lineTo(SIZE, i);
  }
  g.stroke();
  return makePattern(canvas, g);
};

/**
 * Creates a pattern of backslashes slashes.
 * ---------------
 * | \ \ \ \ \ \ |
 * |\ \ \ \ \ \ \|
 * | \ \ \ \ \ \ |
 * |\ \ \ \ \ \ \|
 * ---------------
 * @param {object} style
 * @param {string} style.fill Background color for the fill pattern.
 * @param {string} style.stroke Color for the fill pattern stroke lines.
 */
export const backslash = (style) => {
  const { canvas, g } = makeCanvas();
  g.fillStyle = style.fill;
  g.fillRect(0, 0, SIZE, SIZE);

  g.strokeStyle = style.stroke;
  g.lineWidth = STROKEWIDTH;
  g.beginPath();
  for (let i = 0; i <= SIZE; i += SPACING) {
    g.moveTo(0, SIZE - i);
    g.lineTo(i, SIZE);
    g.moveTo(i, 0);
    g.lineTo(SIZE, SIZE - i);
  }
  g.stroke();
  return makePattern(canvas, g);
};

/**
 * Creates a pattern of crosses.
 * ---------------
 * |+ + + + + + +|
 * |+ + + + + + +|
 * |+ + + + + + +|
 * |+ + + + + + +|
 * ---------------
 * @param {object} style
 * @param {string} style.fill Background color for the fill pattern.
 * @param {string} style.stroke Color for the fill pattern stroke lines.
 */
export const cross = (style) => {
  const { canvas, g } = makeCanvas();
  g.fillStyle = style.fill;
  g.fillRect(0, 0, SIZE, SIZE);

  g.strokeStyle = style.stroke;
  g.lineWidth = STROKEWIDTH;
  g.beginPath();
  for (let px = 0; px <= SIZE; px += SPACING) {
    for (let py = 0; py <= SIZE; py += SPACING) {
      const x = (py / SPACING) % 2 === 0 ? px : px + (SPACING / 2);
      const y = py;
      g.moveTo(x - (SYMBOLSIZE / 2), y);
      g.lineTo(x + (SYMBOLSIZE / 2), y);
      g.moveTo(x, y - (SYMBOLSIZE / 2));
      g.lineTo(x, y + (SYMBOLSIZE / 2));
    }
  }
  g.stroke();
  return makePattern(canvas, g);
};

/**
 * Creates a pattern of diagonal crosses.
 * ---------------
 * |x x x x x x x|
 * |x x x x x x x|
 * |x x x x x x x|
 * |x x x x x x x|
 * ---------------
 * @param {object} style
 * @param {string} style.fill Background color for the fill pattern.
 * @param {string} style.stroke Color for the fill pattern stroke lines.
 */
export const crossDiagonal = (style) => {
  const { canvas, g } = makeCanvas();
  g.fillStyle = style.fill;
  g.fillRect(0, 0, SIZE, SIZE);

  g.strokeStyle = style.stroke;
  g.lineWidth = STROKEWIDTH;
  g.beginPath();
  for (let px = 0; px <= SIZE; px += SPACING) {
    for (let py = 0; py <= SIZE; py += SPACING) {
      const x = (py / SPACING) % 2 === 0 ? px : px + (SPACING / 2);
      const y = py;
      g.moveTo(x - (SYMBOLSIZE / 2), y - (SYMBOLSIZE / 2));
      g.lineTo(x + (SYMBOLSIZE / 2), y + (SYMBOLSIZE / 2));
      g.moveTo(x - (SYMBOLSIZE / 2), y + (SYMBOLSIZE / 2));
      g.lineTo(x + (SYMBOLSIZE / 2), y - (SYMBOLSIZE / 2));
    }
  }
  g.stroke();
  return makePattern(canvas, g);
};

export const table = [
  { id: 'solid', name: 'Solid', renderer: solid },
  { id: 'horiz', name: 'Horizontal lines', renderer: horizontal }, // deprecated
  { id: 'horizontalLines', name: 'Horizontal lines', renderer: horizontal },
  { id: 'vert', name: 'Vertical lines', renderer: vertical }, // deprecated
  { id: 'verticalLines', name: 'Vertical lines', renderer: vertical },
  { id: 'slash', name: 'Slashes', renderer: slash }, // deprecated
  { id: 'slashes', name: 'Slashes', renderer: slash },
  { id: 'backs', name: 'Backslashes', renderer: backslash }, // deprecated
  { id: 'backslashes', name: 'Backslashes', renderer: backslash },
  { id: 'cross', name: 'Crosses', renderer: cross }, // deprecated
  { id: 'crosses', name: 'Crosses', renderer: cross },
  { id: 'diag', name: 'Diagonal crosses', renderer: crossDiagonal }, // deprecated
  { id: 'diagonalCrosses', name: 'Diagonal crosses', renderer: crossDiagonal },
];

/**
 * Gets a fillStyle canvas renderer for the provided pattern ID. See `table` constant above for
 * available patterns.
 *
 * @param {string} id
 * @return {object} { id: string, name: string, renderer: function }
 */
export const getRendererForPattern = (id) => {
  const pattern = table.find(item => item.id === id);
  if (!pattern) {
    throw new Error(`No pattern for id: ${id}`);
  }
  return pattern;
};
