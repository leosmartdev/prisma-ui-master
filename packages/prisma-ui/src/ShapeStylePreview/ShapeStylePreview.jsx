import React from 'react';
import PropTypes from 'prop-types';
import color from 'color';
import { getRendererForPattern } from './build-patterns';

/*
import * as Color from 'lib/color';
import * as Patterns from 'lib/patterns';
*/

export default class ShapeStylePreview extends React.Component {
  static propTypes = {
    /**
     * Width of the preview. In pixels. Default 64px;
     */
    width: PropTypes.number,
    /**
     * Height of the preview. In pixels. Default 64px;
     */
    height: PropTypes.number,
    /**
     * Pattern to fill the polygon with. Defaults to solid color;
     */
    fillPattern: PropTypes.oneOf([
      'solid',
      'horizontalLines',
      'verticalLines',
      'slashes',
      'backslashes',
      'crosses',
      'diagonalCrosses',
    ]),
    /**
     * Color to fill the polygon. Accepts string or object.
     *
     * Eg:
     *  - "rgba(134, 249, 12, 0.5)"
     *  - "#FF9B34"
     *  - { r: 255, g: 255, b: 255 }
     */
    fillColor: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    /**
     * Color of the stroke for the polygon outline. Accepts string or object.
     *
     * Eg:
     *  - "rgba(134, 249, 12, 0.5)"
     *  - "#FF9B34"
     *  - { r: 255, g: 255, b: 255 }
     */
    strokeColor: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    /**
     * When present, outline of the polygon will be hidden.
     */
    hideOutline: PropTypes.bool,
    /**
     * Shape of the style to preview. Renders square or circle.
     */
    shape: PropTypes.oneOf(['polygon', 'circle']),
  }

  static defaultProps = {
    width: 64,
    height: 64,
    fillPattern: 'solid',
    fillColor: '#000000',
    strokeColor: '#000000',
    hideOutline: false,
    shape: 'polygon',
  }

  componentDidMount = () => {
    this.updateCanvas();
  }

  componentDidUpdate = (prevProps) => {
    if (
      this.props.fillColor !== prevProps.fillColor
      || this.props.strokeColor !== prevProps.strokeColor
      || this.props.fillPattern !== prevProps.fillPattern
      || this.props.shape !== prevProps.shape
      || this.props.hideOutline !== prevProps.hideOutline
      || this.props.width !== prevProps.width
      || this.props.height !== prevProps.height
    ) {
      this.updateCanvas();
    }
  }

  setCanvasReference = (canvas) => {
    this.canvas = canvas;
  }


  updateCanvas = () => {
    const {
      width,
      height,
      fillPattern,
      hideOutline,
      shape,
    } = this.props;

    // Guard against canvas ref not being set. Mostly will only see this during testing.
    if (!this.canvas) {
      return;
    }

    const context = this.canvas.getContext('2d');
    const fillColor = color(this.props.fillColor).string();
    const strokeColor = color(this.props.strokeColor).string();

    context.clearRect(0, 0, width, height);
    context.lineWidth = 4;
    context.fillStyle = getRendererForPattern(fillPattern).renderer({
      fill: fillColor,
      stroke: strokeColor,
    });

    if (shape === 'circle') {
      // Draw Circle
      const radius = (width / 2) - 2;
      context.beginPath();
      context.arc(width / 2, height / 2, radius, 0, 2 * Math.PI, false);
      context.fill();

      if (!hideOutline) {
        context.strokeStyle = strokeColor;
        context.stroke();
      }
    } else {
      // Draw Rectangle
      context.fillRect(0, 0, width, height);
      if (!hideOutline) {
        context.strokeStyle = strokeColor;
        context.strokeRect(0, 0, width, height);
      }
    }
  }

  render = () => (
    <canvas
      ref={this.setCanvasReference}
      style={{ verticalAlign: 'middle' }}
      width={this.props.width}
      height={this.props.height}
    />
  )
}
