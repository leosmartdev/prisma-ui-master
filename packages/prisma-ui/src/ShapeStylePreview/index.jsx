/*
 * COPYRIGHT Â© 2018 OROLIA GROUP AND/OR ITS AFFILIATES. ALL RIGHTS ARE STRICTLY RESERVED.
 * THIS SOURCE CODE CONTAINS PROPRIETARY AND CONFIDENTIAL INFORMATION AND DATA AND IS THE SOLE
 * PROPERTY OF OROLIA GROUP AND/OR ITS AFFILIATES.
 * THIS SOURCE CODE MUST NOT BE USED, DISSEMINATED, OR DISTRIBUTED EXCEPT FOR THE AGREED PURPOSE.
 * UNAUTHORIZED USE, REPRODUCTION, OR ISSUE TO ANY THIRD PARTY IS NOT PERMITTED WITHOUT THE PRIOR
 * WRITTEN CONSENT OF THE OROLIA GROUP.
 * THIS SOURCE CODE IS TO BE RETURNED TO THE OROLIA GROUP WHEN THE AGREED PURPOSE IS FULFILLED.
 * ------------------------------------
 *
 * ShapeStylePreview provides a display `<canvas>` to preview a polygon or circle style. The style
 * is used primarily in prisma for map geofences, but can be used for other drawings and diagrams
 * as well.
 *
 * This component provides a preview of that style so a user can see how their changes will
 * affect the look and feel of the shape. You can adjust things like stroke color, fill pattern, and
 * fill color.
 *
 */
import ShapeStylePreview from './ShapeStylePreview';

export default ShapeStylePreview;
