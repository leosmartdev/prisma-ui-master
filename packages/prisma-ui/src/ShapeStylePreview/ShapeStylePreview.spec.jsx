import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import ShapeStylePreview from './ShapeStylePreview';
import * as BuildPatterns from './build-patterns';

jest.mock('./build-patterns');

describe('ShapeStylePreview', () => {
  beforeEach(() => {
    BuildPatterns.getRendererForPattern = jest.fn(() => ({ renderer: jest.fn() }));
  });

  it('mounts', () => {
    const tree = renderer.create(<ShapeStylePreview />);

    expect(tree).toMatchSnapshot();
  });

  describe('properties', () => {
    it('calls for update when fillColor is changed', () => {
      const wrapper = shallow(<ShapeStylePreview />);
      wrapper.instance().updateCanvas = jest.fn();

      wrapper.setProps({ fillColor: '#123456' });

      expect(wrapper.instance().updateCanvas).toHaveBeenCalled();
    });

    it('calls for update when strokeColor is changed', () => {
      const wrapper = shallow(<ShapeStylePreview />);
      wrapper.instance().updateCanvas = jest.fn();

      wrapper.setProps({ strokeColor: '#123456' });

      expect(wrapper.instance().updateCanvas).toHaveBeenCalled();
    });

    it('calls for update when fillPattern is changed', () => {
      const wrapper = shallow(<ShapeStylePreview />);
      wrapper.instance().updateCanvas = jest.fn();

      wrapper.setProps({ fillPattern: 'horizontalLines' });

      expect(wrapper.instance().updateCanvas).toHaveBeenCalled();
    });

    it('calls for update when shape is changed', () => {
      const wrapper = shallow(<ShapeStylePreview />);
      wrapper.instance().updateCanvas = jest.fn();

      wrapper.setProps({ shape: 'circle' });

      expect(wrapper.instance().updateCanvas).toHaveBeenCalled();
    });

    it('calls for update when hideOutline is changed', () => {
      const wrapper = shallow(<ShapeStylePreview />);
      wrapper.instance().updateCanvas = jest.fn();

      wrapper.setProps({ hideOutline: true });

      expect(wrapper.instance().updateCanvas).toHaveBeenCalled();
    });

    it('calls for update when width is changed', () => {
      const wrapper = shallow(<ShapeStylePreview />);
      wrapper.instance().updateCanvas = jest.fn();

      wrapper.setProps({ width: 50 });

      expect(wrapper.instance().updateCanvas).toHaveBeenCalled();
    });

    it('calls for update when height is changed', () => {
      const wrapper = shallow(<ShapeStylePreview />);
      wrapper.instance().updateCanvas = jest.fn();

      wrapper.setProps({ height: 50 });

      expect(wrapper.instance().updateCanvas).toHaveBeenCalled();
    });

    it('does not call for update when another prop is changed', () => {
      const wrapper = shallow(<ShapeStylePreview />);
      wrapper.instance().updateCanvas = jest.fn();

      wrapper.setProps({ foo: 50 });

      expect(wrapper.instance().updateCanvas).not.toHaveBeenCalled();
    });
  });

  describe('updateCanvas', () => {
    let mockContext = {};
    let canvasMock = {};

    beforeEach(() => {
      mockContext = {
        clearRect: jest.fn(),
        beginPath: jest.fn(),
        arc: jest.fn(),
        stroke: jest.fn(),
        fillRect: jest.fn(),
        fillStyle: jest.fn(),
        fill: jest.fn(),
        strokeRect: jest.fn(),
      };

      canvasMock = {
        getContext: jest.fn(() => mockContext),
      };
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('calls correct canvas code when creating a polygon shape', () => {
      const wrapper = shallow(<ShapeStylePreview width={30} height={10} strokeColor="#123456" />);
      wrapper.instance().canvas = canvasMock;

      wrapper.instance().updateCanvas();

      expect(canvasMock.getContext).toHaveBeenCalledWith('2d');
      expect(mockContext.clearRect).toHaveBeenCalledWith(0, 0, 30, 10);
      expect(mockContext.fillRect).toHaveBeenCalledWith(0, 0, 30, 10);
      expect(mockContext.strokeRect).toHaveBeenCalledWith(0, 0, 30, 10);
      expect(mockContext).toMatchObject({
        strokeStyle: 'rgb(18, 52, 86)',
        lineWidth: 4,
        fillStyle: {},
      });
    });

    it('calls correct canvas code when creating a circle shape', () => {
      const wrapper = shallow(<ShapeStylePreview
        width={30}
        height={10}
        strokeColor="#123456"
        shape="circle"
      />);
      wrapper.instance().canvas = canvasMock;

      wrapper.instance().updateCanvas();

      expect(canvasMock.getContext).toHaveBeenCalledWith('2d');
      expect(mockContext.clearRect).toHaveBeenCalledWith(0, 0, 30, 10);
      expect(mockContext.beginPath).toHaveBeenCalled();
      expect(mockContext.arc).toHaveBeenCalledWith(15, 5, 13, 0, Math.PI * 2, false);
      expect(mockContext.fill).toHaveBeenCalled();
      expect(mockContext.stroke).toHaveBeenCalled();
      expect(mockContext).toMatchObject({
        strokeStyle: 'rgb(18, 52, 86)',
        lineWidth: 4,
        fillStyle: {},
      });
    });

    it('doesnt add outline when hideOutline is present in circle mode', () => {
      const wrapper = shallow(<ShapeStylePreview
        width={30}
        height={10}
        strokeColor="#123456"
        shape="circle"
        hideOutline
      />);
      wrapper.instance().canvas = canvasMock;

      wrapper.instance().updateCanvas();

      expect(mockContext.stroke).not.toHaveBeenCalled();
      expect(mockContext.strokeStyle).toBeUndefined();
    });

    it('doesnt add outline when hideOutline is present in polygon mode', () => {
      const wrapper = shallow(<ShapeStylePreview
        width={30}
        height={10}
        strokeColor="#123456"
        shape="polygon"
        hideOutline
      />);
      wrapper.instance().canvas = canvasMock;

      wrapper.instance().updateCanvas();

      expect(mockContext.strokeRect).not.toHaveBeenCalled();
      expect(mockContext.strokeStyle).toBeUndefined();
    });
  });
});
