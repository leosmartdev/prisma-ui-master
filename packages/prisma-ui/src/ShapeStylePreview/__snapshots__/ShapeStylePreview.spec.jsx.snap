// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShapeStylePreview mounts 1`] = `
<canvas
  height={64}
  style={
    Object {
      "verticalAlign": "middle",
    }
  }
  width={64}
/>
`;
